### ISSUE TITLE
connect:packets.ReadPacket block when network is unstable

### ISSUE THREAD

**Issue by foxxnuaa** [eclipse/paho.mqtt.golang#354]
in our project of IOT with edge and server. when network is unstable, we found the packets.ReadPacket blocked.
from log,we can see output "connect started",then no execution.

i think, ReadPacket calls io.ReadFull will block if the server cannot response.
so if we can add some timeout to check or avoid.
orelse we need to put the client connect procedure in the go func().

**Comment by foxxnuaa** [comment]
if we can call SetReadDeadline on c.conn when call packets.ReadPacket?

**Comment by chertov** [comment]
I have the same problem.. https://github.com/eclipse/paho.mqtt.golang/blob/master/packets/packets.go#L105
block on line 105 appears only 3G network
after few seconds i have reconnecting, but messages do not receive!
if i connecting via ethernet it is ok! but via 3G modem i have the block
```
# ./mqttc_armv5te
1 ReadFull start
1 ReadFull end
3 ReadFull start
3 ReadFull end
2 ReadFull start
2 ReadFull end
OnConnect
1 ReadFull start
1 ReadFull end
3 ReadFull start
3 ReadFull end
2 ReadFull start
2 ReadFull end
1 ReadFull start
1 ReadFull end
3 ReadFull start
3 ReadFull end
2 ReadFull start
2 ReadFull end
1 ReadFull start
1 ReadFull end
3 ReadFull start
3 ReadFull end
2 ReadFull start
2 ReadFull end
1 ReadFull start
2020/01/02 05:28:53 client online:  web_jgoozmu85hchpqwgpxodm7
2020/01/02 05:28:55 nals:  3355
... blocked here!
```

**Comment by chertov** [comment]
I found a solution
The problem is on the server side.. i have ubuntu in Digital Ocean cloud
```
sysctl net.ipv4.ip_no_pmtu_disc
* Set it to 1:
sysctl -w net.ipv4.ip_no_pmtu_disc=1
* Check:
sysctl net.ipv4.ip_no_pmtu_disc
* To make it permanent upon reboot
```

**Comment by zm33y** [comment]
We are experiencing the same issue right now. Due to some weird network settings, sometimes TLS data flow fails to start. Thus, the `ReadPacket()` in `connect()` never returns, and the client gets stuck forever without trying to reconnect.

If we set deadline before calling `ReadPacket()` it helps. ReadPacket returns error so the client retries in 5 seconds and eventually succeeds.
```
func (c *client) connect() (byte, bool) {
	DEBUG.Println(NET, "connect started")

	c.conn.SetReadDeadline(time.Now().Add(c.options.ConnectTimeout))
	ca, err := packets.ReadPacket(c.conn)
	c.conn.SetReadDeadline(time.Time{})
...
```
Same applies to `incoming()` function.

**Comment by MattBrittan** [comment]
I believe this was fixed by PR #603 which set a deadline for the handshake. Closing this for now (as it's been a while) but if anyone is still experiencing issues please feel free to reopen.
